AprilSylph/XKit-Rewritten
# Improve loading speed via script file preload

### Description


(rewritten to take further testing into account)

Right now, XKit Rewritten's initial load on a fresh page begins after redpop's hydration finishes and comprises a number of sequential asynchronous tasks. This improves the loading speed of the extension by starting those tasks earlier when they don't depend on the redpop code having executed, reducing the time in which the loaded Tumblr page can be seen without XKit's modifications when the page is navigated to, hard-navigated, or refreshed.

 - The list of installed extensions must be determined by fetching the script index file and querying extension storage; these depend on nothing, so they can be fired at `document_start`.
 - Script and utility files must then be fetched; to avoid running side effects too early and creating a deep nested import waterfall, they are preloaded using `<link rel="preload">` elements, which reliably simultaneously fetch the js files in question from disk without executing them. (This seemed to improve the silent failure to load scripts discussed in #636, but as discussed in the below thread, it does not fully do so and a comprehensive fix for that will be necessary.) This is performed on `interactive`, as per the suggested change.
 - Scripts are then executed once React hydration completes. (The `waitForReactLoaded` change is just a refactor because it took me so long to figure out what it was doing :D)

There are still wait periods post-hydration while scripts that have preferences fetch their json manifests and while scripts' stylesheets are loaded, of course; if we cared about this, having the main script index file indicate whether a script has preferences or a css stylesheet file would allow those to be preloaded, too, but we really don't. (Also, scripts with individual JS files for each function like tweaks could be link preloaded etc etc whatever.)

### Testing steps


~~I don't expect anyone to be able to consistently replicate the problem this solves, as I didn't find a great method, so testing is probably confined to ensuring that XKit continues to function correctly.~~

~~To test the actual fix, merge https://github.com/AprilSylph/XKit-Rewritten/pull/756 into master on a slow PC, enable all of the scripts, and try to observe a failure. Then, merge https://github.com/AprilSylph/XKit-Rewritten/pull/756 into this PR with the same scenario and observe that there are no failures.~~

load it; it go fast
